### DISCUSSION TITLE
Jsonb codec

### DISCUSSION THREAD

**Discussion by go-aegian** [jackc/pgx#1619] [Q&A]
using v5 now the jsonb scan has changed somehow and I need to get help how to scan a marshalled json value into a struct field of type jsonbcodec, below is an example of what I'm trying to do in order to assign to jsonbcodec field a json value, but it does not work, so something for sure I'm doing wrong and need your help understanding the new api.

import "github.com/jackc/pgx/**v5**/pgtype"

type Example struct {
    Field pgtype.JSONBCodec \`json:"field"\`
}

type Custom struct {
    Field1 string \`json:"field1"\`
    Field2 string \`json:"field2"\`
}

a := []Custom {
  {
     Field1: "test1",
     Field2: "test2" ,
  },
}

marshalledA, _:= json.Marshal(a)

example:=&Example{}

m:=pgtype.NewMap()
err:=pgtype.JSONBCodec{}.PlanScan(m, pgtype.JSONBArrayOID, pgx.TextFormatCode, &example.Field).Scan(marshalledA, &example.Field)

fmt.Println(err)

**Comment by jackc** [comment]
You don't use `JSONBCodec` as a scan target. You should scan into the type you want the JSON unmarshalled into or scan into a `[]byte` or `string` to get the raw data.

**Reply by go-aegian** [reply]
So in the example how will it be done?

**Reply by jackc** [reply]
Scan into whatever type of value you actually want -- not a `JSONBCodec`.

**Reply by go-aegian** [reply]
Not sure what do you mean when the jsoncodec field provides planscan, decodevalue but there is no way to assign it back to a []byte variable.

Here is a test example that won't even compile as following what you said

import (
       	"context"
	"encoding/json"
	"fmt"
	"testing" 

        "github.com/jackc/pgx/v5"
	"github.com/jackc/pgx/v5/pgtype"
)

type Example struct {
	Field pgtype.JSONBCodec `json:"field"`
}

type Custom struct {
	Field1 string `json:"field1"`
	Field2 string `json:"field2"`
}

func TestJSONCodec(tst *testing.T) {	
	a := []Custom {
		{
			Field1: "test1",
			Field2: "test2" ,
		},
	}
	jsonBytes, _ := json.Marshal(a)
	
	t := &Example{}
	err := t.Field.PlanScan(pgtype.NewMap(), pgtype.JSONBOID, pgx.TextFormatCode, &t.Field).Scan(jsonBytes, &t.Field)
	fmt.Println(err)
	
	var decoded []byte
	err = t.Field.PlanScan(pgtype.NewMap(), pgtype.JSONBOID, pgx.TextFormatCode, &decoded).Scan(t.Field, &decoded)
	fmt.Println(err)
}

**Reply by jackc** [reply]
You should almost never need to directly use `JSONBCodec` or any other `Codec` for that matter. I'm not sure why you're scanning outside of the context of a database query, but if that's what you really need then you probably want to use https://pkg.go.dev/github.com/jackc/pgx/v5@v5.3.1/pgtype#Map.Scan.
